lib/compiler/src/sys_core_alias.erl

```diff
@@ -0,0 +1,323 @@
+    ordsets:is_disjoint(Vars1, Vars2).
+
+merge_variables(Vars1, Vars2) ->
+    ordsets:union(Vars1, Vars2).
+
+%% get_pattern_keys/2
+%% put_pattern_keys/2
+%%
+%% Gets keys from patterns or add them as aliases.
+
+get_pattern_keys(Patterns) ->
+    lists:foldl(fun get_pattern_keys/2, [], Patterns).
+
+get_pattern_keys(#c_tuple{es=Es}, Acc0) ->
+    Acc1 = accumulate_pattern_keys(tuple, Es, Acc0),
+    lists:foldl(fun get_pattern_keys/2, Acc1, Es);
+get_pattern_keys(#c_cons{hd=Hd,tl=Tl}, Acc0) ->
+    Acc1 = accumulate_pattern_keys(cons, [Hd, Tl], Acc0),
+    get_pattern_keys(Tl, get_pattern_keys(Hd, Acc1));
+get_pattern_keys(#c_alias{pat=Pat}, Acc0) ->
+    get_pattern_keys(Pat, Acc0);
+get_pattern_keys(#c_map{es=Es}, Acc0) ->
+    lists:foldl(fun get_pattern_keys/2, Acc0, Es);
+get_pattern_keys(#c_map_pair{val=Val}, Acc0) ->
+    get_pattern_keys(Val, Acc0);
+get_pattern_keys(_, Acc) ->
+    Acc.
+
+accumulate_pattern_keys(Kind, Nodes, Acc) ->
+    case nodes_to_key(Kind, Nodes) of
+	{ok,Key} -> [Key|Acc];
+	error -> Acc
+    end.
+
+put_pattern_keys(Patterns, []) ->
+    Patterns;
+put_pattern_keys(Patterns, Keys) ->
+    {NewPatterns,Map} =
+	lists:mapfoldl(fun alias_pattern_keys/2, maps:from_list(Keys), Patterns),
+    %% Checks all aliases have been consumed from the map.
+    0 = map_size(Map),
+    NewPatterns.
+
+alias_pattern_keys(#c_tuple{anno=Anno,es=Es0}=Node, Acc0) ->
+    {Es1,Acc1} = lists:mapfoldl(fun alias_pattern_keys/2, Acc0, Es0),
+    nodes_to_alias(tuple, Es0, Anno, Node#c_tuple{es=Es1}, Acc1);
+alias_pattern_keys(#c_cons{anno=Anno,hd=Hd0,tl=Tl0}=Node, Acc0) ->
+    {Hd1,Acc1} = alias_pattern_keys(Hd0, Acc0),
+    {Tl1,Acc2} = alias_pattern_keys(Tl0, Acc1),
+    nodes_to_alias(cons, [Hd0, Tl0], Anno, Node#c_cons{hd=Hd1,tl=Tl1}, Acc2);
+alias_pattern_keys(#c_alias{pat=Pat0}=Node, Acc0) ->
+    {Pat1,Acc1} = alias_pattern_keys(Pat0, Acc0),
+    {Node#c_alias{pat=Pat1}, Acc1};
+alias_pattern_keys(#c_map{es=Es0}=Node, Acc0) ->
+    {Es1,Acc1} = lists:mapfoldl(fun alias_pattern_keys/2, Acc0, Es0),
+    {Node#c_map{es=Es1}, Acc1};
+alias_pattern_keys(#c_map_pair{val=Val0}=Node, Acc0) ->
+    {Val1,Acc1} = alias_pattern_keys(Val0, Acc0),
+    {Node#c_map_pair{val=Val1}, Acc1};
+alias_pattern_keys(Pattern, Acc) ->
+    {Pattern,Acc}.
+
+%% Check if a node must become an alias because
+%% its pattern was used later on as an expression.
+nodes_to_alias(Kind, Inner, Anno, Node, Keys0) ->
+    case nodes_to_key(Kind, Inner) of
+	{ok,Key} ->
+	    case maps:take(Key, Keys0) of
+		{Name,Keys1} ->
+		    Var = cerl:ann_c_var(Anno, Name),
+		    {cerl:ann_c_alias(Anno, Var, Node), Keys1};
+		error ->
+		    {Node,Keys0}
+	    end;
+	error ->
+	    {Node,Keys0}
+     end.
+
+%% Builds the key used to check if a value can be
+%% replaced by an alias. It considers literals,
+%% aliases, variables and cons.
+nodes_to_key(Kind, Nodes) ->
+    case nodes_to_key(Nodes, [], true) of
+	{Acc,false} -> {ok,[Kind|Acc]};
+	{_,true} -> error;
```

Comment:
Alternatively I guess we could say that, if we see a `#c_tuple` with a `#c_literal` inside in both pattern and code, then we should optimize it as it would also reduce the GC pressure, which means that checking for AllLiteral is unnecessary in both cases.